scanner/engine/worker.cpp

```diff
@@ -184,727 +185,737 @@ void analyze_dag(
+      write_profiler_to_file(profiler_output.get(), out_rank, "eval", tag, i,
+                             eval_profilers[pu][1]);
+    }
+    {
+      std::string tag = "post";
+      write_profiler_to_file(profiler_output.get(), out_rank, "eval", tag, i,
+                             eval_profilers[pu][2]);
+    }
+  }
 
-    BACKOFF_FAIL(profiler_output->save());
+  // Save worker profilers
+  u8 save_worker_count = num_save_workers;
+  s_write(profiler_output.get(), save_worker_count);
+  for (i32 i = 0; i < num_save_workers; ++i) {
+    write_profiler_to_file(profiler_output.get(), out_rank, "save", "", i,
+                           save_thread_profilers[i]);
+  }
 
-    VLOG(1) << "Worker " << node_id_ << " finished NewJob";
+  BACKOFF_FAIL(profiler_output->save());
 
-    return grpc::Status::OK;
-  }
+  VLOG(1) << "Worker " << node_id_ << " finished NewJob";
 
-  grpc::Status LoadOp(grpc::ServerContext* context,
-                      const proto::OpInfo* op_info, proto::Empty* empty) {
-    const std::string& so_path = op_info->so_path();
-    void* handle = dlopen(so_path.c_str(), RTLD_NOW | RTLD_LOCAL);
-    LOG_IF(FATAL, handle == nullptr)
-      << "dlopen of " << so_path << " failed: " << dlerror();
-    return grpc::Status::OK;
-  }
+  return grpc::Status::OK;
+}
 
-  grpc::Status Shutdown(grpc::ServerContext* context, const proto::Empty* empty,
-                        Result* result) {
-    trigger_shutdown_.set();
-    result->set_success(true);
-    return grpc::Status::OK;
-  }
+grpc::Status WorkerImpl::LoadOp(grpc::ServerContext* context,
+                                const proto::OpInfo* op_info,
+                                proto::Empty* empty) {
+  const std::string& so_path = op_info->so_path();
+  void* handle = dlopen(so_path.c_str(), RTLD_NOW | RTLD_LOCAL);
+  LOG_IF(FATAL, handle == nullptr)
+    << "dlopen of " << so_path << " failed: " << dlerror();
+  return grpc::Status::OK;
+}
 
- private:
-  std::unique_ptr<proto::Master::Stub> master_;
-  storehouse::StorageConfig* storage_config_;
-  DatabaseParameters db_params_;
-  Flag& trigger_shutdown_;
-  i32 node_id_;
-  storehouse::StorageBackend* storage_;
-  std::map<std::string, TableMetadata*> table_metas_;
-  bool memory_pool_initialized_ = false;
-  MemoryPoolConfig cached_memory_pool_config_;
-};
-
-proto::Worker::Service* get_worker_service(DatabaseParameters& params,
-                                           const std::string& master_address,
-                                           const std::string& worker_port,
-                                           Flag& shutdown) {
-  return new WorkerImpl(params, master_address, worker_port, shutdown);
+grpc::Status WorkerImpl::Shutdown(grpc::ServerContext* context,
+                                  const proto::Empty* empty, Result* result) {
+  trigger_shutdown_.set();
+  result->set_success(true);
+  return grpc::Status::OK;
 }
+
+grpc::Status WorkerImpl::PokeWatchdog(grpc::ServerContext* context,
```

Comment:
Why do workers need their own watchdog? Are we assuming a case where the master would fail (e.g. segfault) without informing the workers?